File: Day3/lesson.c
#include <stdio.h>

// Function prototypes (best practice in C for forward declarations)
int add(int a, int b);
long factorial(int n);

// Main function
int main() {
    // Example 1: Simple function call
    int sum = add(5, 3);
    printf("5 + 3 = %d\n", sum);

    // Example 2: Recursive function
    int num = 5;
    long fact = factorial(num);
    printf("Factorial of %d is %ld\n", num, fact);

    // Example 3: Modular I/O (refactoring Day 1 idea)
    // Note: In C, functions help organize code
    printf("Modular example: ");
    print_hello();

    return 0;
}

// Function definitions
int add(int a, int b) {
    return a + b;  // Simple addition
}

long factorial(int n) {
    if (n <= 1) {
        return 1;  // Base case
    } else {
        return n * factorial(n - 1);  // Recursive call
    }
}

void print_hello() {
    printf("Hello from a function!\n");
}

File: Day3/lesson.py
# This program demonstrates functions and modular programming in Python.
# Functions are defined with def, support various parameter types.
# Best practice: Use type hints and docstrings for clarity.

def add(a: int, b: int) -> int:
    """Return the sum of two integers."""
    return a + b

def factorial(n: int) -> int:
    """Calculate factorial recursively."""
    if n <= 1:
        return 1
    else:
        return n * factorial(n - 1)

def print_hello() -> None:
    """Print a greeting."""
    print("Hello from a function!")

# Main execution
if __name__ == "__main__":  # Best practice: Guard for script execution
    # Example 1: Function call
    sum_result = add(5, 3)
    print(f"5 + 3 = {sum_result}")

    # Example 2: Recursive function
    num = 5
    fact = factorial(num)
    print(f"Factorial of {num} is {fact}")

    # Example 3: Modular call
    print_hello()

    # Note: Python functions can return multiple values, use *args/**kwargs for flexibility
